### ISSUE TITLE
Ternary operator inside a "Where" evaluates all elements even if it's not necessary

### ISSUE THREAD

**Issue by taixes** [dotnet/efcore#20496]
If we have a ternary operator inside a `Where`:

```
var someBoolean = FunctionThatReturnsSomeBoolean();

this.Context
    .Elements
    .Where(e => someBoolean ? e.Id > 5 : true)
    .Any();
```
I'd expect that EF would evaluate the ternary *before* generating the SQL queries. Something like that:

```
var someBoolean = FunctionThatReturnsSomeBoolean();

if (someBoolean)
{
    this.Context
        .Elements
        .Where(e => e.Id)
        .Any();
}
else
{
    this.Context
        .Elements
        .Any();
}
```

This would makes queries simpler and, more important, would avoid bugs like the following one:

```
DateTime? nullableDateTime= FunctionThatReturnsSomeNullableDateTime();

this.Context
    .Elements
    .Where(e => nullableDateTime.HasValue ? e.Date > nullableDateTime.Value : true)
    .Any();
```

In that case, if `nullableDateTime` is null, `nullableDateTime.Value` should never be called, but I'm getting the exception:

`InvalidOperationException: An exception was thrown while attempting to evaluate the LINQ query parameter expression`.

### Steps to reproduce
`Elements` is a table of the context with a Date column.

```
var date = (System.DateTime?)null;

var anyElement = this.Elements
	.Where(e => date.HasValue ? e.Date > date.Value : true)
        .AnyAsync();
```

### Further technical details

EF Core version: .Net Core 3.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: (e.g. .NET Core 3.1)
Operating system: Windows 10
IDE: Visual Studio 2019 16.5.2

**Comment by smitpatel** [comment]
Duplicate of #17942

**Comment by taixes** [comment]
I see it's probably duplicate, sorry. I just wonder if the fix for the `Select` will work also for the `Where`. Probably it will...?

**Comment by smitpatel** [comment]
I am not able to reproduce this on EF Core 3.1.0
I used following code.
```C#
public class Program
    {
        public static async Task Main(string[] args)
        {
            using (var db = new MyContext())
            {
                // Recreate database
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();

                // Seed database


                db.SaveChanges();
            }

            using (var db = new MyContext())
            {
                // Run queries
                var nullableDateTime = FunctionThatReturnsSomeNullableDateTime();
                var query = db.Blogs
                    .Where(b => nullableDateTime.HasValue ? b.Date > nullableDateTime.Value : true)
                    .ToList();
            }
            Console.WriteLine("Program finished.");
        }

        private static DateTime? FunctionThatReturnsSomeNullableDateTime() => null;
    }


    public class MyContext : DbContext
    {
        private static ILoggerFactory ContextLoggerFactory
            => LoggerFactory.Create(b =>
            {
                b
                .AddConsole()
                .AddFilter("", LogLevel.Debug);
            });

        // Declare DBSets
        public DbSet<Blog> Blogs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            // Select 1 provider
            optionsBuilder
                .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=_ModelApp;Trusted_Connection=True;Connect Timeout=5;ConnectRetryCount=0")
                //.UseSqlite("filename=_modelApp.db")
                //.UseInMemoryDatabase(databaseName: "_modelApp")
                //.UseCosmos("https://localhost:8081", @"C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==", "_ModelApp")
                .EnableSensitiveDataLogging()
                .UseLoggerFactory(ContextLoggerFactory);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            // Configure model
        }
    }

    public class Blog
    {
        public int Id { get; set; }
        public DateTime Date { get; set; }
    }
```
Given that the issue was fixed in 3.1.0, please make sure that you are using latest release of EF Core.

**Comment by taixes** [comment]
I am working with 3.1 but I was doing this test in a different branch which was (I don't know why) in 3.0. I upgraded to 3.1 and problem solved. My bad, sorry for wasting your time.
